File: src/java/foundgine-core/src/main/java/com/foundgine/core/semantic/authorization/ISemanticAuthorizationPolicy.java
package com.foundgine.core.semantic.authorization;
import com.foundgine.core.abstractions.*;
/** Provider-independent authorization policy for semantic requests. */
public interface ISemanticAuthorizationPolicy {
    boolean canAccessEntity(EntityId id);
    boolean canAccessField(EntityId entityId, FieldId fieldId);
    boolean canAccessRelationship(EntityId sourceEntityId, RelationshipId relationshipId);
    default boolean canWriteEntity(EntityId id){return false;}
    default boolean canWriteField(EntityId entityId, FieldId fieldId){return false;}
    default boolean canWriteRelationship(EntityId sourceEntityId, RelationshipId relationshipId){return false;}
    default AuthorizationPredicate getPredicate(EntityId id, AuthorizationOperation operation){return null;}
    default AuthorizationDecision getEntityAccess(EntityId id, AuthorizationOperation op){return op==AuthorizationOperation.READ?(canAccessEntity(id)?AuthorizationDecision.ALLOWED:AuthorizationDecision.DENIED):(canWriteEntity(id)?AuthorizationDecision.ALLOWED:AuthorizationDecision.DENIED);}
    default AuthorizationDecision getEntityAccess(EntityId id, AuthorizationOperation op, AuthorizationOperationName name){return getEntityAccess(id,op);}
    default AuthorizationDecision getFieldAccess(EntityId id, FieldId field, AuthorizationOperation op){return op==AuthorizationOperation.READ?(canAccessField(id,field)?AuthorizationDecision.ALLOWED:AuthorizationDecision.DENIED):(canWriteField(id,field)?AuthorizationDecision.ALLOWED:AuthorizationDecision.DENIED);}
    default AuthorizationDecision getRelationshipAccess(EntityId id, RelationshipId relationship, AuthorizationOperation op){return op==AuthorizationOperation.READ?(canAccessRelationship(id,relationship)?AuthorizationDecision.ALLOWED:AuthorizationDecision.DENIED):(canWriteRelationship(id,relationship)?AuthorizationDecision.ALLOWED:AuthorizationDecision.DENIED);}
}

File: src/csharp/Foundgine.Core/Semantic/Authorization/ISemanticAuthorizationPolicy.cs
using Foundgine.Core.Abstractions;

namespace Foundgine.Core.Semantic.Authorization;

/// <summary>
/// Provider-independent authorization policy for semantic requests.
/// Authorization reasons about domain identities only; it must not inspect
/// SQL tables, columns, joins, providers, or transport-specific concepts.
/// </summary>
public interface ISemanticAuthorizationPolicy
{
    // Existing read hooks remain the compatibility surface for simple policies.
    bool CanAccessEntity(EntityId entityId);
    bool CanAccessField(EntityId entityId, FieldId fieldId);
    bool CanAccessRelationship(EntityId sourceEntityId, RelationshipId relationshipId);

    // Write access is deliberately opt-in. Existing policies therefore remain
    // read-only until they explicitly grant writes.
    bool CanWriteEntity(EntityId entityId) => false;
    bool CanWriteField(EntityId entityId, FieldId fieldId) => false;
    bool CanWriteRelationship(EntityId sourceEntityId, RelationshipId relationshipId) => false;

    /// <summary>
    /// Optional provider-independent row/field predicate. The predicate is
    /// preserved in the semantic graph and must not be evaluated away before
    /// provider execution. The default is unrestricted access.
    /// </summary>
    AuthorizationPredicate? GetPredicate(EntityId entityId, AuthorizationOperation operation) => null;

    AuthorizationDecision GetEntityAccess(EntityId entityId, AuthorizationOperation operation) =>
        operation == AuthorizationOperation.Read
            ? (CanAccessEntity(entityId) ? AuthorizationDecision.Allowed : AuthorizationDecision.Denied)
            : (CanWriteEntity(entityId) ? AuthorizationDecision.Allowed : AuthorizationDecision.Denied);

    /// <summary>
    /// Named-operation refinement of <see cref="GetEntityAccess(EntityId, AuthorizationOperation)"/>.
    /// The default falls back to the coarse Read/Write decision above, so
    /// existing policies keep their current behavior unchanged. Override this
    /// only when a policy needs to distinguish domain-specific write intents
    /// (for example "Invoice.Pay" versus "Invoice.Update") that the coarse
    /// <see cref="AuthorizationOperation.Write"/> gate does not separate.
    /// A denial here must never be weaker than the coarse decision would be:
    /// this refinement may only narrow access, never widen it.
    /// </summary>
    AuthorizationDecision GetEntityAccess(
        EntityId entityId,
        AuthorizationOperation operation,
        AuthorizationOperationName? name) =>
        GetEntityAccess(entityId, operation);

    AuthorizationDecision GetFieldAccess(EntityId entityId, FieldId fieldId, AuthorizationOperation operation) =>
        operation == AuthorizationOperation.Read
            ? (CanAccessField(entityId, fieldId) ? AuthorizationDecision.Allowed : AuthorizationDecision.Denied)
            : (CanWriteField(entityId, fieldId) ? AuthorizationDecision.Allowed : AuthorizationDecision.Denied);

    AuthorizationDecision GetRelationshipAccess(
        EntityId sourceEntityId,
        RelationshipId relationshipId,
        AuthorizationOperation operation) =>
        operation == AuthorizationOperation.Read
            ? (CanAccessRelationship(sourceEntityId, relationshipId)
                ? AuthorizationDecision.Allowed
                : AuthorizationDecision.Denied)
            : (CanWriteRelationship(sourceEntityId, relationshipId)
                ? AuthorizationDecision.Allowed
                : AuthorizationDecision.Denied);
}